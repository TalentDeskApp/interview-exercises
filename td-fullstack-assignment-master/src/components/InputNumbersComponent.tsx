import React from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';
import { FormValues } from '@/types';

const minNumbers = 3;
const maxNumbers = 1000;

const validationRules = {
  required: 'This field is required',
  validate: {
    isNumberFormat: (value: string) => {
      const numberPattern = /^(\d+)(,\d+)*$/;
      return (
        numberPattern.test(value) || 'Enter valid numbers separated by commas'
      );
    },
    hasValidNumberCount: (value: string) => {
      const values = value.split(',').filter(Boolean);
      const numberCount = values.length;
      return numberCount >= minNumbers && numberCount <= maxNumbers
        ? true
        : `Enter between ${minNumbers} and ${maxNumbers} numbers`;
    },
  },
};

// TODO to improve UX add loading state, and clear button
export const InputNumbersComponent = ({
  onSubmit,
}: {
  onSubmit: SubmitHandler<FormValues>;
}) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>();

  return (
    <div className="mt-8 w-full max-w-md">
      <form
        onSubmit={handleSubmit(onSubmit)}
        className="bg-white p-6 rounded-lg shadow-lg w-full max-w-md"
      >
        <div className="mb-4 flex items-center space-x-2">
          <input
            {...register('numbers', validationRules)}
            id="inputField"
            data-testid="input-field"
            type="text"
            className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
            placeholder="Enter numbers, e.g. 1,2,3"
          />
          <button
            className="top-2 right-2 px-4 py-2 bg-blue-500 text-white font-semibold rounded-md hover:bg-blue-600 focus:outline-none focus:ring-2 focus:ring-blue-500"
            data-testid="submit-button"
          >
            Enter
          </button>
        </div>

        <div className="h-3">
          {errors.numbers && (
            <p className="mt-2 text-red-500 text-sm" data-testid="error-field">
              {errors.numbers.message}
            </p>
          )}
        </div>
      </form>
    </div>
  );
};
