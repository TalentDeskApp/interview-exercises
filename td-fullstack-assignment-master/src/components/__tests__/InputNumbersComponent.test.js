import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { InputNumbersComponent } from '../InputNumbersComponent';

describe('<InputNumbersComponent />', () => {
  it('Renders InputNumbersComponent', () => {
    const { asFragment } = render(<InputNumbersComponent />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Form throws error if enter less than 3 numbers', async () => {
    const mockOnSubmit = jest.fn();

    render(<InputNumbersComponent onSubmit={mockOnSubmit} />);

    const inputField = screen.getByTestId('input-field');
    const submitButton = screen.getByTestId('submit-button');

    await act(async () => {
      fireEvent.change(inputField, { target: { value: '123' } });
      fireEvent.click(submitButton);
    });

    const errorField = await screen.findByTestId('error-field');
    expect(errorField.textContent).toEqual('Enter between 3 and 1000 numbers');
  });

  it('Form throws error if enter more than 1000 numbers', async () => {
    const mockOnSubmit = jest.fn();

    render(<InputNumbersComponent onSubmit={mockOnSubmit} />);

    const inputField = screen.getByTestId('input-field');
    const submitButton = screen.getByTestId('submit-button');

    const value = Array.from({ length: 1001 }, (_, index) => index + 1).join(
        ',',
    );

    await act(async () => {
      fireEvent.change(inputField, { target: { value } });
      fireEvent.click(submitButton);
    });

    const errorField = await screen.findByTestId('error-field');
    expect(errorField.textContent).toEqual('Enter between 3 and 1000 numbers');
  });

  it('Form throws error if input is empty', async () => {
    const mockOnSubmit = jest.fn();

    render(<InputNumbersComponent onSubmit={mockOnSubmit} />);

    const submitButton = screen.getByTestId('submit-button');

    await act(async () => {
      fireEvent.click(submitButton);
    });

    const errorField = await screen.findByTestId('error-field');
    expect(errorField.textContent).toEqual('This field is required');
  });

  it('Form throws error if use wrong characters', async () => {
    const mockOnSubmit = jest.fn();

    render(<InputNumbersComponent onSubmit={mockOnSubmit} />);

    const inputField = screen.getByTestId('input-field');
    const submitButton = screen.getByTestId('submit-button');

    await act(async () => {
      fireEvent.change(inputField, { target: { value: '1.2.3' } });
      fireEvent.click(submitButton);
    });

    const errorField = await screen.findByTestId('error-field');
    expect(errorField.textContent).toEqual(
        'Enter valid numbers separated by commas',
    );
  });

  it('Successfully accept entered numbers', async () => {
    const mockOnSubmit = jest.fn();

    render(<InputNumbersComponent onSubmit={mockOnSubmit} />);

    const inputField = screen.getByTestId('input-field');
    const submitButton = screen.getByTestId('submit-button');
    const value = '1,2,3';

    await act(async () => {
      fireEvent.change(inputField, { target: { value } });
      fireEvent.click(submitButton);
    });

    const errorField = screen.queryByTestId('error-field');
    expect(errorField).toBeNull();

    const updatedInputField = screen.getByTestId('input-field');
    const inputValue = updatedInputField.value;
    expect(inputValue).toBe(value);
  });
});