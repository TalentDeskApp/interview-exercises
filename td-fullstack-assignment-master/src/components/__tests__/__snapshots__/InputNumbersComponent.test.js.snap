// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputNumbersComponent /> Renders InputNumbersComponent 1`] = `
<DocumentFragment>
  <div
    class="mt-8 w-full max-w-md"
  >
    <form
      class="bg-white p-6 rounded-lg shadow-lg w-full max-w-md"
    >
      <div
        class="mb-4 flex items-center space-x-2"
      >
        <input
          class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
          data-testid="input-field"
          id="inputField"
          name="numbers"
          placeholder="Enter numbers, e.g. 1,2,3"
          type="text"
        />
        <button
          class="top-2 right-2 px-4 py-2 bg-blue-500 text-white font-semibold rounded-md hover:bg-blue-600 focus:outline-none focus:ring-2 focus:ring-blue-500"
          data-testid="submit-button"
        >
          Enter
        </button>
      </div>
      <div
        class="h-3"
      />
    </form>
  </div>
</DocumentFragment>
`;
